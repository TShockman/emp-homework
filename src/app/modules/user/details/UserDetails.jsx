import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import Loading from '../../common/loading/Loading';
import OrderList from '../../order/orderList';
import style from './userDetails.scss';
import cx from 'classnames';

export default class UserDetails extends PureComponent {
  static propTypes = {
    router: PropTypes.object.isRequired,
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      email: PropTypes.string
    }),
    requestUser: PropTypes.func.isRequired,
    id: PropTypes.number,
    push: PropTypes.func.isRequired
  };

  componentDidMount = () => {
    const {requestUser, router, id, push} = this.props;
    const {id: routeId} = router.params;
    if (id === null) {
      return push('/login');
    } else if (id.toString() !== routeId) {
      // naive privileges check: all users only can view their own page
      return push(`/user/${id}`);
    } else {
      return requestUser({id: router.params.id});
    }
  };

  render() {
    let contents;
    if (!this.props.user) {
      contents = <Loading/>;
    } else {
      const {firstName, lastName, email} = this.props.user;
      contents = (
          <div className={cx("card", "welcomeCard")}>
            <h1>Welcome {firstName} {lastName}!</h1>
            <h2>{email}</h2>
          </div>
      );
    }

    return (
        <div className={cx("userDetails", "container-fluid")}>
          <div className="row">
            <div className="col">
              {contents}
            </div>
          </div>
          <OrderList/>
        </div>
    );
  }
}