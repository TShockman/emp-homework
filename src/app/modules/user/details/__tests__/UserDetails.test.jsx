/* eslint-disable import/first */
jest.mock('../../../order/orderList');

import React from 'react';
import UserDetails from '../UserDetails';
import {mount} from 'enzyme';
import ReactTestRenderer from 'react-test-renderer/shallow';

const renderer = new ReactTestRenderer();

const getProps = () => ({
  router: {params: {id: '1'}},
  user: {
    id: 1,
    firstName: 'Bob',
    lastName: 'Jones',
    email: 'bob.jones.1@mail.net'
  },
  requestUser: jest.fn(),
  id: 1,
  push: jest.fn()
});

describe('User Details', () => {
  it('Renders loading without user', () => {
    const props = getProps();
    props.user = null;
    expect(renderer.render(<UserDetails {...props}/>)).toMatchSnapshot();
  });
  it('Renders consistently with user', () => {
    const props = getProps();
    expect(renderer.render(<UserDetails {...props}/>)).toMatchSnapshot();
  });
  it('Redirects to login on mount if not logged in', () => {
    const props = getProps();
    props.id = null;
    const comp = <UserDetails {...props}/>;
    expect(props.push).not.toHaveBeenCalled();
    expect(props.requestUser).not.toHaveBeenCalled();

    const wrapper = mount(comp);
    expect(props.push).toHaveBeenLastCalledWith('/login');
    expect(props.requestUser).not.toHaveBeenCalled();
    wrapper.unmount();
  });
  it('Redirects to own users page if user tries to access another user page', () => {
    const props = getProps();
    props.id = 777;
    const comp = <UserDetails {...props}/>;
    expect(props.push).not.toHaveBeenCalled();
    expect(props.requestUser).not.toHaveBeenCalled();

    const wrapper = mount(comp);
    expect(props.push).toHaveBeenLastCalledWith('/user/777');
    expect(props.requestUser).not.toHaveBeenCalled();
    wrapper.unmount();
  });
  it('Fetches user information on mount if logged in', () => {
    const props = getProps();
    const comp = <UserDetails {...props}/>;
    expect(props.push).not.toHaveBeenCalled();
    expect(props.requestUser).not.toHaveBeenCalled();

    const wrapper = mount(comp);
    expect(props.push).not.toHaveBeenCalled();
    expect(props.requestUser).toHaveBeenLastCalledWith({id: '1'});
    wrapper.unmount();
  });
});