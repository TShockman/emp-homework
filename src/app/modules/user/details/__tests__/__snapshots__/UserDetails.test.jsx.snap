// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Details Renders consistently with user 1`] = `
<div
  className="userDetails container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="card welcomeCard"
      >
        <h1>
          Welcome 
          Bob
           
          Jones
          !
        </h1>
        <h2>
          bob.jones.1@mail.net
        </h2>
      </div>
    </div>
  </div>
  <Connect(OrderList) />
</div>
`;

exports[`User Details Renders loading without user 1`] = `
<div
  className="userDetails container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <Loading />
    </div>
  </div>
  <Connect(OrderList) />
</div>
`;
